### PR TITLE
Bug 1880902: automatically set the dnsPolicy with ClusterFirstWithHostNet when endpoint publishing strategy type is set to hostnetwork

### PR DESCRIPTION
**Pull request by miheer**
Bug 1880902 - If the endpoint publishing strategy type is set to hostnetwork then
we should automatically set the dnsPolicy with ClusterFirstWithHostNet.

### CODE REVIEW

File: pkg/operator/controller/ingress/deployment.go [hunk 1]
```diff
@@ -386,6 +386,7 @@ func desiredRouterDeployment(ci *operatorv1.IngressController, ingressController
 		// problems or firewall restrictions.
 		deployment.Spec.Template.Spec.Containers[0].LivenessProbe.Handler.HTTPGet.Host = "localhost"
 		deployment.Spec.Template.Spec.Containers[0].ReadinessProbe.Handler.HTTPGet.Host = "localhost"
+		deployment.Spec.Template.Spec.DNSPolicy = "ClusterFirstWithHostNet"
```
Comment (sgreene570) [inline pkg/operator/controller/ingress/deployment.go:389]:
```suggestion
		deployment.Spec.Template.Spec.DNSPolicy = corev1.DNSClusterFirstWithHostNet
```
Probably better if we just use the constant directly from the API definitions, right?

**Comment by Miciah** [comment]
We should add a unit test in `TestDeploymentConfigChanged`.  Also, we probably should add `deployment.Spec.Template.Spec.DNSPolicy = corev1.DNSClusterFirst` for the non-HostNetworking case so we always set an explicit policy, which will prevent the comparison in `deploymentConfigChanged` from getting confused when the API sets a default value.

**Comment by miheer** [comment]
@Miciah PTAL

File: pkg/operator/controller/ingress/deployment_test.go [hunk 2]
```diff
@@ -901,6 +930,7 @@ func TestDeploymentConfigChanged(t *testing.T) {
 						},
 					},
 					Spec: corev1.PodSpec{
+						DNSPolicy: corev1.DNSClusterFirstWithHostNet,
```
Comment (sgreene570) [inline pkg/operator/controller/ingress/deployment_test.go:933]: Would it make more sense to set the test deployment's `DNSPolicy` to `corev1.DNSClusterFirst` since that's the API defaulted value (and then mutate that value to `corev1.DNSClusterFirstWithHostNet` in the new test cases)?

**Review by sgreene570 — COMMENTED** [review]
some small comments

**Comment by miheer** [comment]
After the changes -> Unit tests are passing

[miheer@miheer cluster-ingress-operator]$ GO111MODULE=on go test -v -run TestDesiredRouterDeployment ./pkg/operator/controller/ingress/
=== RUN   TestDesiredRouterDeployment
--- PASS: TestDesiredRouterDeployment (0.01s)
PASS
ok  	github.com/openshift/cluster-ingress-operator/pkg/operator/controller/ingress	0.017s
[miheer@miheer cluster-ingress-operator]$ GO111MODULE=on go test -v -run TestDeploymentConfigChanged ./pkg/operator/controller/ingress/
=== RUN   TestDeploymentConfigChanged
--- PASS: TestDeploymentConfigChanged (0.01s)
PASS
ok  	github.com/openshift/cluster-ingress-operator/pkg/operator/controller/ingress	(cached)
[miheer@miheer cluster-ingress-operator]$

File: pkg/operator/controller/ingress/deployment_test.go [hunk 3]
```diff
@@ -875,6 +887,21 @@ func TestDeploymentConfigChanged(t *testing.T) {
...
+		{
+			description: "if dnsPolicy is changed when ",
+			mutate: func(deployment *appsv1.Deployment) {
+				deployment.Spec.Template.Spec.DNSPolicy = corev1.DNSClusterFirst
+			},
+			expect: false,
+		},
+
+		{
+			description: "if dnsPolicy is not changed",
```
[the remark this answers is not in the document]
Comment (miheer) [reply pkg/operator/controller/ingress/deployment_test.go:899]: oh yes I will add them...just switched the code.

[the remark this answers is not in the document]
Comment (miheer) [reply pkg/operator/controller/ingress/deployment_test.go]: @sgreene570 done

File: pkg/operator/controller/ingress/deployment_test.go [hunk 4]
```diff
@@ -229,6 +229,10 @@ func TestDesiredRouterDeployment(t *testing.T) {
 		t.Error("expected host network to be false")
 	}
 
+	if deployment.Spec.Template.Spec.DNSPolicy != corev1.DNSClusterFirst {
+		t.Errorf("expected dnsPolicy to be %q, got %q", corev1.DNSClusterFirst, deployment.Spec.Template.Spec.DNSPolicy)
```
Comment (sgreene570) [inline pkg/operator/controller/ingress/deployment_test.go:233]:
```suggestion
		t.Errorf("expected dnsPolicy to be %s, got %s", corev1.DNSClusterFirst, deployment.Spec.Template.Spec.DNSPolicy)
```

File: pkg/operator/controller/ingress/deployment_test.go [hunk 5]
```diff
@@ -307,6 +311,9 @@ func TestDesiredRouterDeployment(t *testing.T) {
 	if deployment.Spec.Template.Spec.HostNetwork != false {
 		t.Error("expected host network to be false")
 	}
+	if deployment.Spec.Template.Spec.DNSPolicy != corev1.DNSClusterFirst {
+		t.Errorf("expected dnsPolicy to be #{corev1.DNSClusterFirst}")
```
Comment (sgreene570) [inline pkg/operator/controller/ingress/deployment_test.go:315]:
```suggestion
		t.Errorf("expected dnsPolicy to be %s, got %s", corev1.DNSClusterFirst, deployment.Spec.Template.Spec.DNSPolicy)
```

File: pkg/operator/controller/ingress/deployment_test.go [hunk 6]
```diff
@@ -459,6 +466,11 @@ func TestDesiredRouterDeployment(t *testing.T) {
 	if deployment.Spec.Template.Spec.HostNetwork != true {
 		t.Error("expected host network to be true")
 	}
+
+	if deployment.Spec.Template.Spec.DNSPolicy != corev1.DNSClusterFirstWithHostNet {
+		t.Errorf("expected dnsPolicy to be #{corev1.DNSClusterFirstWithHostNet}")
```
Comment (sgreene570) [inline pkg/operator/controller/ingress/deployment_test.go:471]:
```suggestion
		t.Errorf("expected dnsPolicy to be %s, got %s", corev1.DNSClusterFirstWithHostNet, deployment.Spec.Template.Spec.DNSPolicy)
```

File: pkg/operator/controller/ingress/deployment_test.go [hunk 7]
```diff
@@ -875,6 +887,21 @@ func TestDeploymentConfigChanged(t *testing.T) {
 			},
 			expect: true,
 		},
+		{
+			description: "if dnsPolicy is not changed",
```
Comment (sgreene570) [inline pkg/operator/controller/ingress/deployment_test.go:891]: I think this case is covered by the "if nothing changes" case.

File: pkg/operator/controller/ingress/deployment_test.go [hunk 8]
```diff
@@ -305,7 +309,10 @@ func TestDesiredRouterDeployment(t *testing.T) {
 	}
 	checkRollingUpdateParams(t, deployment, intstr.FromString("25%"), intstr.FromString("25%"))
 	if deployment.Spec.Template.Spec.HostNetwork != false {
-		t.Error("expected host network to be false")
+		t.Errorf("expected dnsPolicy to be %s, got %s", corev1.DNSClusterFirst, deployment.Spec.Template.Spec.DNSPolicy)
```
Comment (Miciah) [inline pkg/operator/controller/ingress/deployment_test.go:312]: Please revert this change.

**Comment by Miciah** [comment]
/retitle Bug 1880902: automatically set the dnsPolicy with ClusterFirstWithHostNet when endpoint publishing strategy type is set to hostnetwork
/lgtm
Thanks!
